### DISCUSSION TITLE
mainmenu.cpp

### DISCUSSION THREAD

**Discussion by JoBergeron** [diasurgical/DevilutionX#2392] [General]
Why do we have two files named "mainmenu.cpp" ? It seems to be causing issues with Xcode when generating the corresponding .o files (it's also sort of unusual to have two files named the same within the same project)

**Comment by AJenbo** [comment]
originally DiabloUI and the application used to be two separate projects.

**Reply by JoBergeron** [reply]
Not sure if there's a compiler flag to fix this or not. I'll be doing a bit of research. Are you open to the idea to add a prefix to files from the DiabloUI folder maybe?

**Reply by AJenbo** [reply]
I'm a bit surprised that Xcode needs filenames to be globally unique, what if there was a header in a dependency that was also called mainmenu?

Rather then a prefix i would prefer a descriptive name.

**Reply by JoBergeron** [reply]
Headers would be alright. It's the .cpp files that get turned into .o files that seem to be the problem. Maybe it's the way cmake generated the xcode project file, I don't know. I'm more experienced working with visual studio and gcc so this is all a learning experience for me.

**Reply by JoBergeron** [reply]
VS seems seems to get around this by conserving the folder structure 
![image](https://user-images.githubusercontent.com/1339414/125950811-86275596-fd29-44bd-b470-f29fbfc7f41f.png)

maybe there's a way flag for LLVM to do the same. I'll poke around

**Reply by AJenbo** [reply]
Sounds odd that it it should be a problem since we already build for Mac and Android via LLVM

**Reply by JoBergeron** [reply]
I've already compiled the game on Mac with success using the makefile approach. However, when I generate an xcode project using cmake, I run into this problem which is very odd.

**Reply by AJenbo** [reply]
Let me know if you have a new name suggestion or figure out how to solve the issue with Xcode :)

**Reply by JoBergeron** [reply]
The issue in Xcode seems to go a little like this:

- All cpp files that are compiled have their object files in the same repository

- In order to differentiate between some .o files that are named the same, xcode appends a unique identifier to the filename

ex: the two mainmenu.o files get renamed to mainmenu-12343456ABF1234.o and mainmenu-EF3434863B2F232.o

- In Cmake, we've made an Object style library (Instead of a Static or Dynamic lib), so in the linker options, you will get a flag telling DevilutionX to link with mainmenu.o, which doesn't exist)

I have no idea how to fix this, I spent a lot of time looking at it and cannot figure it out. What I propose it to rename Source/mainmenu.cpp (and its .h while we're there) to Source/ingamemainmenu.cpp/.h. This is by far the simplest solution I've found. 

Another solution I found was to create another Object library for the DiabloUI stuff. This also worked but sort of sucked since we had to add all the dependencies to it again.

So yeah, Source/ingamemainmenu.cpp/.h is the name I propose

**Reply by AJenbo** [reply]
ingamemainmenu.cpp doesn't work as a name as it is not the ingame menu but the part that calls up the main menu and reacts to selections in it.

**Reply by JoBergeron** [reply]
I hadn't really looked at it's content to be honest. I was sort of guessing there.

Maybe have one called mainmenuimpl?

**Reply by AJenbo** [reply]
`impl` is that short for implementation?

I would rather just call it `menu.cpp`. The ingame one is called `gamemenu`/`gmenu`

**Reply by JoBergeron** [reply]
menu.cpp seems alright yeah

**Reply by AJenbo** [reply]
Ok, feel free to send the PR for that :)
